crates/webidl/src/util.rs

```diff
@@ -693,6 +638,149 @@ impl<'src> FirstPassRecord<'src> {
             }
         }
     }
+
+    pub fn create_imports(
+        &self,
+        kind: backend::ast::ImportFunctionKind,
+        id: &OperationId<'src>,
+        data: &OperationData2<'src>,
+    )
+        -> Vec<backend::ast::ImportFunction>
+    {
+        // First up expand all the signatures in `data` into all signatures that
+        // we're going to generate. These signatures will be used to determine
+        // the names for all the various functions.
+        #[derive(Clone)]
+        struct ExpandedSig<'a> {
+            orig: &'a Signature<'a>,
+            args: Vec<IdlType<'a>>,
+        }
+
+        let mut actual_signatures = Vec::new();
+        'outer:
+        for signature in data.signatures.iter() {
+            let start = actual_signatures.len();
+
+            // Start off with an empty signature, this'll handle zero-argument
+            // cases and otherwise the loop below will continue to add on to this.
+            actual_signatures.push(ExpandedSig {
+                orig: signature,
+                args: Vec::with_capacity(signature.args.len()),
+            });
+
+            for (i, arg) in signature.args.iter().enumerate() {
+                let cur = actual_signatures.len();
+                // If any argument in this signature can't be converted we have
+                // to throw out the entire signature, so revert back to the
+                // beginning and then keep going.
```

Comment:
Implementing the filtering out of unsupported arg types is making the logic harder to follow than necessary -- especially with the backtracking via truncate. Can we do a `filter_map` before we even start the `'outer` loop so that we aren't doing so much all at once?